packages/openapi-v3-types/src/openapi-v3-spec-types.ts

```diff
@@ -0,0 +1,401 @@
+
+export type OpenApiSpec = OAS3.OpenAPIObject;
+/**
+ * Custom extensions can use arbitrary type as the value,
+ * e.g. a string, an object or an array.
+ */
+// tslint:disable-next-line:no-any
+export type ExtensionValue = any;
+
+/**
+ * The location of a parameter.
+ * Possible values are "query", "header", "path" or "cookie".
+ * Specification:
+ * https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md
+ * search "The location of the parameter"
+ */
+export type ParameterLocation = 'query' | 'header' | 'path' | 'cookie';
+
+/**
+ * The style of a parameter.
+ * Describes how the parameter value will be serialized.
+ * (serialization is not implemented yet)
+ * Specification:
+ * https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#responseObject
+ */
+export type ParameterStyle =
+  | 'matrix'
+  | 'label'
+  | 'form'
+  | 'simple'
+  | 'spaceDelimited'
+  | 'pipeDelimited'
+  | 'deepObject';
+
+/**
+ * The Schema Object allows the definition of input and output data types.
+ * the properties consiste of two parts:
+ * - taken directly from the JSON Schema, is described by interface `JSONType`
+ * - taken from the JSON Schema, but definitions were adjusted to the
+ *   OpenAPI Specification, is described by interface `OAS3SchemaObject`
+ */
+export interface SchemaObject extends JSONType, OAS3SchemaObject {}
+
+/**
+ * Part of OpenAPI Schema Object, The following properties are taken from the
+ * JSON Schema definition but their definitions were adjusted to the OpenAPI
+ * Specification.
+ */
+export interface OAS3SchemaObject extends ISpecificationExtension {
+  nullable?: boolean;
+  discriminator?: DiscriminatorObject;
+  readOnly?: boolean;
+  writeOnly?: boolean;
+  xml?: XMLObject;
+  externalDocs?: ExternalDocumentationObject;
+  example?: ExtensionValue;
+  examples?: ExtensionValue[];
+  deprecated?: boolean;
+
+  type?: string;
+  allOf?: (SchemaObject | ReferenceObject)[];
+  oneOf?: (SchemaObject | ReferenceObject)[];
+  anyOf?: (SchemaObject | ReferenceObject)[];
+  not?: SchemaObject | ReferenceObject;
+  items?: SchemaObject | ReferenceObject;
+  properties?: {[propertyName: string]: SchemaObject | ReferenceObject};
+  additionalProperties?: SchemaObject | ReferenceObject;
+  description?: string;
+  format?: string;
+  default?: ExtensionValue;
+}
+
+/**
+ * JSON type - This is part of the Schema object.
+ * The following properties are taken directly from the JSON Schema
+ * definition and follow the same specifications.
+ * See https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#schema-object
+ */
+
+export type JSONType = {
+  title?: string;
+  multipleOf?: number;
+  maximum?: number;
+  exclusiveMaximum?: number;
+  minimum?: number;
+  exclusiveMinimum?: number;
+  maxLength?: number;
+  minLength?: number;
+  // (This string SHOULD be a valid regular expression, according to the ECMA 262 regular expression dialect)
+  pattern?: string;
```

Comment:
[RegExp is a class in JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp), so we should have some sort of typing for the compiler's sake that lets users know when they're not utilizing this correctly. I would create a union type of `RegExp | string` or something similar and use it here.

Reply:
thanks @kjdelisle, the reason I put it as `string` is the swagger2 spec also treats it as a `string`, see [old spec](https://github.com/strongloop/loopback-next/blob/master/packages/openapi-spec/src/openapi-spec-v2.ts#L674-L677)

And you can find that comment from OpenAPI document: https://github.com/strongloop/loopback-next/blob/master/packages/openapi-spec/src/openapi-spec-v2.ts#L674-L677

Not an excuse of not investigating it 😀 ...but I am avoiding digging into these trivial nested properties in this story, I know they worth investigation, I can give you ten more examples, but I still feel it's out of scope, sorry.

This is also why I said I don't like doing things in a rush, sometimes I know I can do better but I also know the time limit, have to leave them as they are.

Take the `pattern` as an example, there is probably a good reason to use `string` here, probably not, maybe `RegExp` in JS is the right way to go. Whichever reality it is, I need time to prove.